crates/telio-proxy/src/proxy.rs

```diff
@@ -259,10 +331,35 @@ impl Runtime for StateEgress {
             return Self::sleep_forever().await;
         }
 
-        // Outbound data (from telio to WG)
-        if let Some((pk, msg)) = self.input.recv().await {
-            if let (Some(socket), Some(wg_addr)) = (self.sockets.get(&pk), self.wg_addr) {
-                let _ = socket.send_to(msg.get_payload(), wg_addr).await;
+        self.cleanup_mutes();
+        let dummy_sleep = &mut PinnedSleep::new(Duration::MAX, PublicKey::default());
+        let mute_futures = {
+            if self.mutes.is_empty() {
+                vec![dummy_sleep]
+            } else {
+                self.mutes
+                    .iter_mut()
+                    .map(|(_, mute)| mute)
+                    .collect::<Vec<_>>()
+            }
+        };
+
+        tokio::select! {
+            // Outbound data (from telio to WG)
+            Some((pk, msg)) = self.input.recv() => {
+                if let (Some((socket, muted)), Some(wg_addr)) = (self.sockets.get(&pk), self.wg_addr) {
+                    if !*muted {
+                        let _ = socket.send_to(msg.get_payload(), wg_addr).await;
+                    } else {
+                        telio_log_debug!("Peer {:?} is muted", pk);
+                    }
+                }
+            },
+            (pk, _, _) = select_all(mute_futures) => {
+                if let Some((_, muted)) = self.sockets.get_mut(&pk) {
+                    telio_log_debug!("Unmuting peer: {:?}", pk);
+                    *muted = false;
+                }
             }
```

Comment:
I somewhat feel that you should just implement a check when the packet is about to send, meaning that there should be no `select_all()` but a simple check just before the `socket.send_to()` if the peer is muted or even better there should be no `muted` hash map but as hash map with `PublickKey -> Instant (last time muted)` and you should check here if the peer was muted more than 15s ago, then send. It would be much simpler to implement and more reliable IMHO

Reply:
I partially agree, that simplifies a lot if we know the default sleep would never change, in the future if we accept custom mute durations it will add complexity. So I would prefer to just wait for the mute sleeps to expire